raft/raft.go

```diff
@@ -1226,12 +1314,22 @@ func (r *raft) removeNode(id uint64) {
 
 func (r *raft) resetPendingConf() { r.pendingConf = false }
 
-func (r *raft) setProgress(id, match, next uint64) {
+func (r *raft) setProgress(id, match, next uint64, isLearner bool) {
+	if isLearner {
+		if _, ok := r.prs[id]; ok {
+			panic(fmt.Sprintf("%x can't change voter to learner for %x", r.id, id))
+		}
+		r.learnerPrs[id] = &Progress{Next: next, Match: match, ins: newInflights(r.maxInflight), IsLearner: true}
+		return
+	}
+
+	delete(r.learnerPrs, id)
 	r.prs[id] = &Progress{Next: next, Match: match, ins: newInflights(r.maxInflight)}
```

Comment:
i would handle not learner case first, since it is easier. so it improves readability.